apps/src/storage/dataBrowser/DataVisualizer.jsx

```diff
@@ -0,0 +1,135 @@
+    // from redux state
+    tableColumns: PropTypes.arrayOf(PropTypes.string).isRequired,
+    tableName: PropTypes.string.isRequired,
+    // "if all of the keys are integers, and more than half of the keys between 0 and
+    // the maximum key in the object have non-empty values, then Firebase will render
+    // it as an array."
+    // https://firebase.googleblog.com/2014/04/best-practices-arrays-in-firebase.html
+    tableRecords: PropTypes.oneOfType([PropTypes.object, PropTypes.array])
+      .isRequired
+  };
+
+  state = {...INITIAL_STATE};
+
+  handleClose = () => {
+    this.setState({isVisualizerOpen: false});
+  };
+
+  handleChange = (field, value) => {
+    this.setState({[field]: value});
+  };
+
+  render() {
+    const dropdownField = (fieldName, displayName, values) => (
+      <div id={fieldName} style={rowStyle.container}>
+        <label style={rowStyle.description}>{displayName}</label>
+        <select
+          value={this.state[fieldName]}
+          onChange={event => this.handleChange(fieldName, event.target.value)}
+        >
+          <option value="">Select</option>
+          {values.map(value => (
+            <option key={value} value={value}>
+              {value}
+            </option>
+          ))}
+        </select>
+      </div>
+    );
+
+    const modalBody = (
+      <div>
+        <h1> Explore {this.props.tableName} </h1>
+        <h2> Overview </h2>
+        <div id="selection-area">
+          <div id="chartTitleRow" style={rowStyle.container}>
+            <label style={rowStyle.description}>Chart Title</label>
+            <input
+              style={rowStyle.input}
+              onChange={event =>
+                this.handleChange('chartTitle', event.target.value)
+              }
+            />
+          </div>
+
+          {dropdownField('chartType', 'Chart Type', [
+            'Bar Chart',
+            'Histogram',
+            'Cross Tab',
+            'Scatter Plot'
+          ])}
+
+          {this.state.chartType === 'Histogram' && (
+            <div id="numBinsRow" style={rowStyle.container}>
+              <label style={rowStyle.description}>Bins</label>
+              <input
+                style={rowStyle.input}
+                value={this.state.numBins}
+                onChange={event =>
+                  this.handleChange('numBins', event.target.value)
+                }
+              />
+            </div>
+          )}
+          {(this.state.chartType === 'Bar Chart' ||
+            this.state.chartType === 'Histogram') &&
+            dropdownField('values', 'Values', this.props.tableColumns)}
+
+          {(this.state.chartType === 'Cross Tab' ||
+            this.state.chartType === 'Scatter Plot') && (
+            <div>
+              {dropdownField('xValues', 'X Values', this.props.tableColumns)}
+              {dropdownField('yValues', 'Y Values', this.props.tableColumns)}
+            </div>
+          )}
+        </div>
+        <div id="chart-area" />
```

Comment:
What's the purpose of this id? In general, we try to avoid ids in react components because components are intended to be reusable.

Reply:
Offline: It's needed for rendering a chart.
✔️